### ISSUE TITLE
Finish new engine

### ISSUE THREAD

**Issue by psFried** [psFried/flo#24]
Brain dump of everything required to finish the new engine and replace the old one.

Features yet to be implemented in the new engine:

- [x] basic produce and consume
- [x] test for filtering events based on namespace
- [x] consume from multiple partitions at once
- [x] Consumer batch operations
- [x] event retention based on time
- [x] Consumer stopping at an arbitrary point
- [x] Remove old engine code
- [x] Allow protocol to be generic over event types
- [ ] logging of performance metrics

Features to be done in the client:

- [x] basic async client (building block for everything else)
- [ ] specialized producer for produce-only work
- [x] ~~specialized consumer for producing responses~~ (not so much needed anymore since dealing with Iterators and Streams is so much more ergonomic)
- [x] port over sync client to be based on async client (should also get the cli/bench tools working with the new engine)

Once this stuff is done, we should be able to just wipe out all the old engine code and then get back to work on new features.

**Comment by psFried** [comment]
Closing this issue, as I've decided to defer both of the remaining pieces until later. I created a separate issue for collecting performance metrics, #36. As for the specialized producer, I think that was just a little premature, since the design of the producer would have to change significantly once it has to communicate with a cluster instead of a single server.
